laplace/baselaplace.py

```diff
@@ -1003,12 +1047,505 @@ def functional_variance(self, Js: torch.Tensor) -> torch.Tensor:
         samples = samples * self.posterior_scale.reshape(1, self.n_params)
-        return self.mean.reshape(1, self.n_params) + samples
+        return self.map_estimate.reshape(1, self.n_params) + samples
 
 
 class FunctionalLaplace(BaseLaplace):
-    pass
+    """
+    Applying the GGN (General Gauss Newton) approximation for the Hessian in the Laplace approximation of the posterior
+    turns the underlying probabilistic model from a BNN into a GLM (generalized linear model).
+    This GLM (in the weight space) is equivalent to a GP (in the function space), see
+    [Approximate Inference Turns Deep Networks into Gaussian Processes (Khan et al., 2019)](https://arxiv.org/abs/1906.01930)
+
+    This class implements the (approximate) GP inference through which
+    we obtain the desired quantities (posterior predictive, marginal log-likelihood).
+    See [Improving predictions of Bayesian neural nets via local linearization (Immer et al., 2021)](https://arxiv.org/abs/2008.08400)
+    for more details.
+
+    Note that for `likelihood='classification'`, we approximate \( L_{NN} \\) with a diagonal matrix
+    ( \\( L_{NN} \\) is a block-diagonal matrix, where blocks represent Hessians of per-data-point log-likelihood w.r.t.
+     neural network output \\( f \\), See Appendix [A.2.1](https://arxiv.org/abs/2008.08400) for exact definition). We
+     resort to such an approximation because of the (possible) errors found in Laplace approximation for
+     multiclass GP classification in Chapter 3.5 of [R&W 2006 GP book](http://www.gaussianprocess.org/gpml/),
+     see the question
+     [here](https://stats.stackexchange.com/questions/555183/gaussian-processes-multi-class-laplace-approximation)
+     for more details. Alternatively, one could also resort to *one-vs-one* or *one-vs-rest* implementations
+     for multiclass classification, however, that is not (yet) supported here.
+
+    Parameters
+    ----------
+    M : int
+        number of data points for Subset-of-Data (SOD) approximate GP inference.
+        By default (`M=None`), all data points from train dataset are used
+    diagonal_kernel : bool
+        GP kernel here is product of Jacobians, which results in a \\( C \\times C\\) matrix where \\(C\\) is the output
+        dimension. If `diagonal_kernel=True`, only a diagonal of a GP kernel is used. This is (somewhat) equivalent to
+        assuming independent GPs across output channels.
+
+    See `BaseLaplace` class for the full interface.
+    """
+    # key to map to correct subclass of BaseLaplace, (subset of weights, Hessian structure)
+    _key = ('all', 'GP')
+
+    def __init__(self, model, likelihood, M=None, sigma_noise=1., prior_precision=1.,
+                 prior_mean=0., temperature=1., backend=BackPackGGN, backend_kwargs=None,
+                 diagonal_kernel=False):
+        assert backend in [BackPackGGN, AsdlGGN]
+        if isinstance(backend, AsdlGGN):
+            assert likelihood == 'classification'
```

Comment:
I think it would be consistent not to raise this error here but let it get raised by the backend call later as in the other laplace classes. The reason is that it is the responsibility of Asdl to know if it can implement regression (which it will be able to, soon) and hence this check might lead to unexpected behaviour once it can.